File: DataEngineering/05_Stream_Processing/projects/python/05_kafka_to_db_consumer/main.py
import json
import logging
from kafka import KafkaConsumer
from sqlalchemy import create_engine, Column, Integer, String, Float, DateTime
from sqlalchemy.orm import declarative_base, sessionmaker
from datetime import datetime

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Configuration
KAFKA_BROKER = "localhost:9092"
KAFKA_TOPIC = "user_activity"
DB_CONNECTION_STRING = "postgresql://user:password@localhost:5432/analytics_db"

Base = declarative_base()

class UserActivity(Base):
    __tablename__ = 'user_activities'
    
    id = Column(Integer, primary_key=True, autoincrement=True)
    user_id = Column(String, index=True)
    activity_type = Column(String)
    value = Column(Float)
    timestamp = Column(DateTime, default=datetime.utcnow)

def get_db_session():
    engine = create_engine(DB_CONNECTION_STRING)
    Base.metadata.create_all(engine)
    Session = sessionmaker(bind=engine)
    return Session()

def consume_and_store():
    logger.info("Starting Kafka to DB Consumer...")
    
    try:
        consumer = KafkaConsumer(
            KAFKA_TOPIC,
            bootstrap_servers=[KAFKA_BROKER],
            auto_offset_reset='earliest',
            enable_auto_commit=True,
            group_id='db_consumer_group',
            value_deserializer=lambda x: json.loads(x.decode('utf-8'))
        )
    except Exception as e:
        logger.error(f"Failed to connect to Kafka: {e}")
        return

    session = get_db_session()
    
    logger.info(f"Listening for messages on '{KAFKA_TOPIC}'...")
    
    # Process messages in batches to optimize DB inserts
    batch_size = 100
    batch = []
    
    try:
        for message in consumer:
            data = message.value
            
            try:
                activity = UserActivity(
                    user_id=data.get('user_id'),
                    activity_type=data.get('activity_type'),
                    value=float(data.get('value', 0.0)),
                    timestamp=datetime.fromisoformat(data.get('timestamp', datetime.utcnow().isoformat()))
                )
                batch.append(activity)
            except Exception as e:
                logger.error(f"Error parsing message data: {e}")
                continue
            
            # Commit batch when size is reached
            if len(batch) >= batch_size:
                try:
                    session.bulk_save_objects(batch)
                    session.commit()
                    logger.info(f"Committed {len(batch)} records to database.")
                    batch.clear()
                except Exception as e:
                    session.rollback()
                    logger.error(f"Database commit failed: {e}")
                    
    except KeyboardInterrupt:
        logger.info("Stopping consumer...")
    finally:
        # Commit remaining
        if batch:
            try:
                session.bulk_save_objects(batch)
                session.commit()
                logger.info(f"Committed final {len(batch)} records to database.")
            except Exception as e:
                logger.error(f"Final DB commit failed: {e}")
        
        consumer.close()
        session.close()

if __name__ == "__main__":
    consume_and_store()

File: DataEngineering/05_Stream_Processing/projects/golang/05_kafka_to_db_consumer/main.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/segmentio/kafka-go"
	"gorm.io/driver/postgres"
	"gorm.io/gorm"
	"gorm.io/gorm/logger"
)

var (
	kafkaBroker = "localhost:9092"
	kafkaTopic  = "user_activity"
	dbDSN       = "host=localhost user=user password=password dbname=analytics_db port=5432 sslmode=disable"
)

type UserActivity struct {
	ID           uint      `gorm:"primaryKey"`
	UserID       string    `gorm:"index"`
	ActivityType string
	Value        float64
	Timestamp    time.Time
}

type ActivityMessage struct {
	UserID       string  `json:"user_id"`
	ActivityType string  `json:"activity_type"`
	Value        float64 `json:"value"`
	Timestamp    string  `json:"timestamp"`
}

func main() {
	fmt.Println("Starting Kafka to DB Consumer...")

	// Initialize Database connection
	db, err := gorm.Open(postgres.Open(dbDSN), &gorm.Config{
		Logger: logger.Default.LogMode(logger.Error),
	})
	if err != nil {
		log.Fatalf("Failed to connect to database: %v", err)
	}

	// Auto Migrate the schema
	if err := db.AutoMigrate(&UserActivity{}); err != nil {
		log.Fatalf("Failed to migrate database: %v", err)
	}

	ctx, cancel := context.WithCancel(context.Background())
	defer cancel()

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
	go func() {
		<-sigChan
		log.Println("Shutting down gracefully...")
		cancel()
	}()

	reader := kafka.NewReader(kafka.ReaderConfig{
		Brokers:  []string{kafkaBroker},
		Topic:    kafkaTopic,
		GroupID:  "db-consumer-group",
		MinBytes: 10e3,
		MaxBytes: 10e6,
	})
	defer reader.Close()

	log.Printf("Listening for messages on topic '%s'...", kafkaTopic)

	batchSize := 100
	var batch []UserActivity

	for {
		m, err := reader.ReadMessage(ctx)
		if err != nil {
			if ctx.Err() != nil {
				break
			}
			log.Printf("Error reading message: %v", err)
			continue
		}

		var msg ActivityMessage
		if err := json.Unmarshal(m.Value, &msg); err != nil {
			log.Printf("Error unmarshalling message: %v", err)
			continue
		}

		ts, err := time.Parse(time.RFC3339, msg.Timestamp)
		if err != nil {
			ts = time.Now()
		}

		activity := UserActivity{
			UserID:       msg.UserID,
			ActivityType: msg.ActivityType,
			Value:        msg.Value,
			Timestamp:    ts,
		}

		batch = append(batch, activity)

		if len(batch) >= batchSize {
			if err := db.Create(&batch).Error; err != nil {
				log.Printf("Error inserting batch to DB: %v", err)
			} else {
				log.Printf("Committed %d records to database.", len(batch))
			}
			batch = batch[:0] // Clear batch
		}
	}

	// Flush remaining batch
	if len(batch) > 0 {
		if err := db.Create(&batch).Error; err != nil {
			log.Printf("Error inserting final batch to DB: %v", err)
		} else {
			log.Printf("Committed final %d records to database.", len(batch))
		}
	}

	log.Println("Kafka to DB Consumer stopped.")
}
